### ISSUE TITLE
[Regression] TextBox with ContextMenu: Right click clears selection

### ISSUE THREAD

**Issue by Deadpikle** [AvaloniaUI/Avalonia#3674]
_Previous issue_: #2692

### Dev environment

* Latest `master` (36c34b4 as of writing)

### Description

When you right click a text box and the text box has a `ContextMenu`, it clears the selection. This is a problem for context menus (copy, paste, etc.).

### The problem

When the `ContextMenu` opens, it opens a `Popup`. The `ContextMenu` takes focus when the Popup opens (https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/ContextMenu.cs#L162), which then causes `TextBox.OnLostFocus` to be called.

```
protected override void OnLostFocus(RoutedEventArgs e)
{
    base.OnLostFocus(e);
    SelectionStart = 0;
    SelectionEnd = 0;
    _presenter?.HideCaret();
}
```

This function doesn't care why the `TextBox` lost focus -- the selection is removed. We need the `ContextMenu` to be focused so that keyboard navigation works (4a5e11f6aaee79a3183aa7d703b60c4fa912a971).

The naive fix of "hey is the thing that got focused our `ContextMenu`? yes? ok let's not clear our selection." within `OnLostFocus` isn't a good fix because if you open the context menu and then select a _different_ `TextBox`, the original one will never have `OnLostFocus` called, so the selection is never cleared.

**Comment by grokys** [comment]
Yeah I've noticed this too. We need to take a look at WPF and see how this is handled there.

**Comment by grokys** [comment]
Fixed by #4498
